### PR TITLE
Contact blobid: Always return 'blobId' in /set{create.update} and  /get responses.

### PR DESCRIPTION
**Pull request by ksmurchison**
Matching PR for Cassandane exists

### CODE REVIEW

File: changes/next/contact_blobid [hunk 1]
```diff
@@ -0,0 +1,13 @@
+Description:
+
+Always return 'blobId' in /set{create.update} and  /get responses.
```
Comment (rsto) [inline changes/next/contact_blobid:3]: I'd state the JMAP object type that this change applies to. e.g. "Always return 'blobId' in in JMAP Contact /set...."

File: imap/jmap_contact.c [hunk 2]
```diff
@@ -1031,9 +1032,9 @@ static void _contacts_set(struct jmap_req *req, unsigned kind)
         struct entryattlist *annots = NULL;
         strarray_t *flags = NULL;
 
+        json_t *item = json_pack("{}");
```
Comment (rsto) [inline imap/jmap_contact.c:1035]: `json_pack("{}") ` can be simplified in this case to `json_object()`

**Review by rsto — COMMENTED** [review]
Generally looks good to me. Just found two nits.

**Review by rsto — APPROVED** [review]
Looks good to me

**Comment by ksmurchison** [comment]
@rsto One last review of the final two commits please.  Forgot to use the generic blob fetcher for email parts.

File: imap/jmap_api.h [hunk 3]
```diff
@@ -171,11 +171,19 @@ typedef struct {
...
     const char *accept_mime;     // input to the handler
+    unsigned decode : 1;         // input to the handler
     struct buf blob;             // output from the handler
-    const char *content_type;    // output from the handler
+    char *content_type;          // output from the handler
+    char *encoding;              // output from the handler
     const char *errstr;          // output from the handler
 } jmap_getblob_context_t;
 
+#define GETBLOB_CTX_INITIALIZER(from_accountid, blobid, accept_mime, decode) \
```
Comment (rsto) [inline imap/jmap_api.h:181]: This deviates from the currently used idioms in the codebase: None of the current INITIALIZER macros takes arguments. Instead they initialize a struct with default values that make the data structure ready for use without further initialization. (And the way the macro currently is defined is not safe with regards to potential duplicate evaluation).

For parameterized initializers there's a couple of foo_init functions, but unfortunately they do not all follow the same pattern. Some are void functions, others return an initialized struct. I tend to use void functions.

For the particular case, though, I would just not use an INITIALIZER at all. Is the blob context initialized at more than one code section?

File: imap/jmap_api.h [hunk 4]
```diff
@@ -171,11 +171,19 @@ typedef struct {
...
+    { from_accountid, blobid, accept_mime, decode,                           \
+            BUF_INITIALIZER, NULL, NULL, NULL };
+
+void jmap_getblob_ctx_free(jmap_getblob_context_t *ctx);
```
Comment (rsto) [inline imap/jmap_api.h:185]: This should read `_fini` rather than `_free`.

**Review by rsto — CHANGES_REQUESTED** [review]
Found two idiomatic issues that I think should be addressed.

File: imap/jmap_api.h (hunk 3 above)
Comment (ksmurchison) [reply imap/jmap_api.h]: Fixed in latest commit

**Pull request merged**